### PR TITLE
Add support for precision for TIMESTAMP in Postgresql type mapping

### PR DESCRIPTION
**Pull request by losipiuk**
Replaces https://github.com/prestosql/presto/pull/4570

### CODE REVIEW

**Comment by mosabua** [comment]
Please add type mapping docs for postgresql as part of this PR so it is clear to users.

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/PostgreSqlQueryRunner.java [hunk 1]
```diff
@@ -49,11 +49,23 @@ public static DistributedQueryRunner createPostgreSqlQueryRunner(
             Map<String, String> connectorProperties,
             Iterable<TpchTable<?>> tables)
             throws Exception
+    {
+        return createPostgreSqlQueryRunner(server, extraProperties, connectorProperties, tables, 3);
+    }
+
+    public static DistributedQueryRunner createPostgreSqlQueryRunner(
```
Comment (findepi) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/PostgreSqlQueryRunner.java:56]: I would rather drop this commit.
The added burden of an overload is not justified for arguably small benefit this brings.

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 2]
```diff
@@ -1199,15 +1200,6 @@ public void testArrayTimestampWithTimeZone(boolean insertWithPresto)
         tests.execute(getQueryRunner(), sessionWithArrayAsArray(), dataSetup);
     }
 
-    @DataProvider
-    public Object[][] testTimestampWithTimeZoneDataProvider()
```
Comment (findepi) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1203]: The cmt title is

> Rename misleading testTimestampWithTimeZoneDataProvider

it's removal not "rename"

"misleading" is misleading. It was very nicely called "<test-name>DataProvider", i do not find anything misleading in that and next time would call it the same.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java [hunk 3]
```diff
@@ -382,24 +397,34 @@ public static LongReadFunction timestampReadFunction()
...
     }
 
-    public static long toPrestoTimestamp(LocalDateTime localDateTime)
+    public static long toPrestoTimestamp(TimestampType timestampType, LocalDateTime localDateTime)
     {
-        return localDateTime.atZone(UTC).toInstant().toEpochMilli() * MICROSECONDS_PER_MILLISECOND;
+        long precision = timestampType.getPrecision();
+        checkArgument(precision <= MAX_SHORT_PRECISION, "Precision is out of range: %s", precision);
+        Instant instant = localDateTime.atZone(UTC).toInstant();
+        return instant.getEpochSecond() * MICROSECONDS_PER_SECOND + roundDiv(instant.getNano(), NANOSECONDS_PER_MICROSECOND);
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:417]: if timestampType.precision < 6, you need to round/truncate here

@martint what is the correct way to do this?
there are some utility methods in `io.prestosql.spi.type.Timestamps`, but I do not see something that's immediately applicable.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java [hunk 4]
```diff
@@ -382,24 +397,34 @@ public static LongReadFunction timestampReadFunction()
...
     }
 
-    public static LocalDateTime fromPrestoTimestamp(long value)
+    public static LocalDateTime fromPrestoTimestamp(TimestampType timestampType, long epochMicros)
     {
-        return Instant.ofEpochMilli(floorDiv(value, MICROSECONDS_PER_MILLISECOND)).atZone(UTC).toLocalDateTime();
+        int precision = timestampType.getPrecision();
+        checkArgument(precision <= MAX_SHORT_PRECISION && precision >= 0, "Precision is out of range: %s", precision);
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:423]: You do not really need to validate this here. If someone provided `epochMicros`, it;'s all you need.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java [hunk 5]
```diff
@@ -382,24 +397,34 @@ public static LongReadFunction timestampReadFunction()
...
+        long epochSecond = floorDiv(epochMicros, MICROSECONDS_PER_SECOND);
+        int nanoFraction = floorMod(epochMicros, MICROSECONDS_PER_SECOND) * NANOSECONDS_PER_MICROSECOND;
+        Instant instant = Instant.ofEpochSecond(epochSecond, nanoFraction);
+        return LocalDateTime.ofInstant(instant, UTC);
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:427]: we could unify & reuse with `io.prestosql.spi.type.Timestamps#formatTimestamp(int precision, long epochMicros, int picosOfMicro)`

File: presto-mysql/src/main/java/io/prestosql/plugin/mysql/MySqlClient.java [hunk 6]
```diff
@@ -262,7 +262,7 @@ public WriteMapping toWriteMapping(ConnectorSession session, Type type)
         }
         if (TIMESTAMP_MILLIS.equals(type)) {
             // TODO use `timestampWriteFunction`
-            return WriteMapping.longMapping("datetime", timestampWriteFunctionUsingSqlTimestamp());
+            return WriteMapping.longMapping("datetime", timestampWriteFunctionUsingSqlTimestamp(TIMESTAMP));
```
Comment (findepi) [inline presto-mysql/src/main/java/io/prestosql/plugin/mysql/MySqlClient.java:265]:
```suggestion
            return WriteMapping.longMapping("datetime", timestampWriteFunctionUsingSqlTimestamp(TIMESTAMP_MILLIS));
```

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 7]
```diff
@@ -1005,24 +1005,38 @@ public void testTime(boolean insertWithPresto, ZoneId sessionZone)
...
-                .addRoundTrip(timestampDataType(), timeDoubledInJvmZone)
-                .addRoundTrip(timestampDataType(), timeDoubledInVilnius)
-                .addRoundTrip(timestampDataType(), epoch) // epoch also is a gap in JVM zone
-                .addRoundTrip(timestampDataType(), timeGapInJvmZone1)
-                .addRoundTrip(timestampDataType(), timeGapInJvmZone2)
-                .addRoundTrip(timestampDataType(), timeGapInVilnius)
-                .addRoundTrip(timestampDataType(), timeGapInKathmandu);
+        DataTypeTest tests = DataTypeTest.create(true);
+        for (int precision : List.of(3, 6)) {
+            // test all standard cases with precision 3 and 6 to make sure the long and short TIMESTAMP WITH TIME ZONE
```
Comment (findepi) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1010]: "TIMESTAMP WITH TIME ZONE"

also, is 3 and 6 actually different cases?
i think in timestamp case, we go with the same code path

(ok to leave the test coverage, but if true, change the commnt)

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 8]
```diff
@@ -1032,32 +1046,50 @@ public void testTimestamp(boolean insertWithPresto, ZoneId sessionZone)
...
-        tests.execute(getQueryRunner(), session, dataSetup);
+        if (insertWithPresto) {
+            tests.execute(getQueryRunner(), session, prestoCreateAsSelect(sessionWithArrayAsArray(), "test_array_timestamp"));
+        }
+        else {
+            tests.execute(getQueryRunner(), session, postgresCreateAndInsert("tpch.test_array_timestamp"));
+        }
+    }
+
+    private DataType<List<LocalDateTime>> arrayOfTimestampDataType(int precision, boolean insertWithPresto)
```
Comment (findepi) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1085]: this belong to `Simplify flow in timestamp with timezone mapping tests` commit?

**Review by findepi — APPROVED** [review]

**Comment by losipiuk** [comment]
> Please add type mapping docs for postgresql as part of this PR so it is clear to users.

This is a point change regarding one specific type and adding general type mapping documentation is substantial effort itself. 
Added https://github.com/prestosql/presto/issues/5131 to track that.

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java (hunk 2 above)
Comment (losipiuk) [reply presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1203]: Changed to "Use generic trueFalse data provider for timestamp tests"

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java (hunk 5 above)
Comment (losipiuk) [reply presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:427]: We could but I am not sure if we could extend spi with random methods. We should keep it fairly stable. Also the spi methods cares about `picosOfMicro` which we do not care about here. With `picosOfMicro` you cannot create a method which just returns `LocalDateTime`, which we need here. And have a specific method version which does not take `picosOfSeconds` and returns `LocalDateTime` seems to specific to me, to be put in spi.

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java (hunk 8 above)
Comment (losipiuk) [reply presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1085]: no. This is specific to `TIMESTAMP` not `TIMESTAMP w/TZ`. So it belongs to last commit.

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 9]
```diff
@@ -1032,32 +1046,50 @@ public void testTimestamp(boolean insertWithPresto, ZoneId sessionZone)
...
-        DataSetup dataSetup;
-        if (insertWithPresto) {
-            dataType = arrayDataType(timestampDataType());
-            dataSetup = prestoCreateAsSelect(sessionWithArrayAsArray(), "test_array_timestamp");
-        }
-        else {
-            dataType = arrayDataType(timestampDataType(), "timestamp[]");
-            dataSetup = postgresCreateAndInsert("tpch.test_array_timestamp");
+        DataTypeTest tests = DataTypeTest.create(true);
+        for (int precision : List.of(3, 6)) {
```
Comment (losipiuk) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1050]: I will drop this for. As representation is same for 3 and 6 for TIMESTAMP

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java (hunk 7 above)
Comment (losipiuk) [reply presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:1010]: Oh - you noticed it too - yeah - it same case. I dropped the coverage.

**Pull request merged**